### ISSUE TITLE
testthat messes up sorting when dealing with different case types

### ISSUE THREAD

**Issue by sumedh10** [r-lib/testthat#1001]
- The minimum reprex needed to demonstrate this bug is a simple package, which can be found [here](https://github.com/sumedh10/test)
- The bug can _only_ be reproduced when running `Cmd + Shift + T` or `devtools::test()`, not by running the individual test file 
- In fact, things work as expected when run in the console _and_ as an individual test file  
- Finally, I was able to reproduce this bug _only_ when dealing with _lower case vs upper case_ letters

This is how `base::sort()` works in R: 

```r
> sort(c("B", "b"))
[1] "b" "B"
```

Similarly, 

```r
> library(dplyr)
> df <- data.frame(test_col = c("B", "b"), stringsAsFactors = FALSE)
> print(df)
  test_col
1        B
2        b
> print(arrange(df, test_col))
  test_col
1        b
2        B
```

**tl;dr**

All this is to say that lower case `"b"` is supposed to be _less than_ upper case `"B"`. 

However, in the context of `tests/testthat/test-*`, `base::sort()` and `dplyr::arrange()` don't seem to work as expected (FWIW, `base::order()` and `>` don't seem to work too). However, `stringr::str_sort()` does seem to work fine. 

### Steps to Reproduce 

1. Clone the [repository](https://github.com/sumedh10/test) 
2. Run the tests in `tests/testthat/test-hello.R` either using `Cmd + Shift + T` or `devtools::test()` (not as an individual test file) 
3. You'll notice that two tests fail (only the test for `stringr::str_sort()` passes) -- that's because both `sort()` and `arrange()` don't sort the vector and column as expected

> NOTE: The closest somewhat related comment I was able to dig up is [this](https://github.com/r-lib/testthat/issues/341#issuecomment-171324520), though not sure if it relates to the current issue.

**Comment by hadley** [comment]
I think this is https://github.com/tidyverse/dplyr/issues/2751, which is fixed in dev dplyr.
